File: vokimi_api/Src/dtos/responses/view_test_page/discussions/TestDiscussionCommentVm.cs
using vokimi_api.Src.db_related.db_entities.tests_related.discussions;
using vokimi_api.Src.db_related.db_entities_ids;

namespace vokimi_api.Src.dtos.responses.view_test_page.discussions
{
    public record class TestDiscussionCommentVm(
        string CommentId,
        string AuthorId,
        string AuthorUsername,
        string AuthorProfilePicture,
        string Text,//IDiscussionCommentContent or attachment
        int VotesRating,
        int TotalVotesCount,
        string CreatedAtDateTime,
        bool? IsViewersVoteUp,
        TestDiscussionCommentVm[] ChildVms
    )
    {
        public static TestDiscussionCommentVm FromComment(
            TestDiscussionsComment comment,
            Dictionary<TestDiscussionsCommentId, bool> viewersVotes
        ) => new(
            comment.Id.Value.ToString(),
            comment.AuthorId.Value.ToString(),
            comment.Author.Username,
            comment.Author.ProfilePicturePath,
            comment.Text,
            CalculateVotesRating(comment.CommentVotes),
            comment.CommentVotes.Count,
            comment.CreatedAt.ToString("HH:mm dd.MM.yyyy"),
            viewersVotes.TryGetValue(comment.Id, out var val) ? val : null,
            comment.ChildComments.Select((c) => FromComment(c, viewersVotes)).ToArray()
        );
        private static int CalculateVotesRating(ICollection<DiscussionsCommentVote> votes) =>
            votes.Sum(v => v.IsUp ? 1 : -1);
    }
}

File: vokimi_frontend/src/ts/page_classes/view_test_page_classes/middle_section_tabs_classes/discussions_tab_classes/TestDiscussionCommentVm.ts
export class TestDiscussionCommentVm {
    public readonly commentId: string;
    public readonly authorId: string;
    public readonly authorUsername: string;
    public readonly authorProfilePicture: string;
    public readonly text: string;
    public readonly votesRating: number;
    public readonly totalVotesCount: number;
    public readonly createdAtDateTime: string;
    public readonly isViewersVoteUp: boolean | null;
    public childVms: TestDiscussionCommentVm[];

    constructor(
        commentId: string,
        authorId: string,
        authorUsername: string,
        authorProfilePicture: string,
        text: string,
        votesRating: number,
        totalVotesCount: number,
        createdAtDateTime: string,
        isViewersVoteUp: boolean | null,
        childVms: TestDiscussionCommentVm[]
    ) {
        this.commentId = commentId;
        this.authorId = authorId;
        this.authorUsername = authorUsername;
        this.authorProfilePicture = authorProfilePicture;
        this.text = text;
        this.votesRating = votesRating;
        this.totalVotesCount = totalVotesCount;
        this.createdAtDateTime = createdAtDateTime;
        this.isViewersVoteUp = isViewersVoteUp;
        this.childVms = childVms;
    }
}